### DISCUSSION TITLE
Carbon Copy Newsletter #1

### DISCUSSION THREAD

**Discussion by wolffg** [carbon-language/carbon-lang#3684] [Announcements]
# **Carbon Copy, Feb. 2024**

Welcome to **Carbon Copy**, a periodic update on the Carbon language! 

We know that there are folks excited about the Carbon language who may lack time to follow the day-to-day work on Discord or Github. **Carbon Copy** is designed for those people who want a high-level view of what's happening on the project. This issue is a little longer than future ones as we're starting from scratch.

If you'd like to subscribe, you can join [announce@carbon-lang.dev](https://groups.google.com/a/carbon-lang.dev/g/announce). **Carbon Copy** should arrive roughly every other month.

## **Spotlight: Carbon goals and roadmap for 2024**

As this is our first ever edition, and the first **Carbon Copy** of 2024, let's look back at 2023. 

To borrow from the 2023 [retrospective](https://github.com/carbon-language/carbon-lang/blob/e5d5acc035c6e077e4bf63c557ea6626890372f8/proposals/p3564.md#retrospective-on-2023), our [roadmap for 2023](https://github.com/carbon-language/carbon-lang/blob/840cb1bed7cf9bd57e000cb4a61e986c383d3038/docs/project/roadmap.md) focused on getting both the language and tools ready for evaluation.

We made fantastic progress on getting the language (design) ready by closing the most critical design gaps identified at the beginning of the year. While the tooling side made remarkable progress, it also started at a much earlier stage and so has a lot further to go. 

We accepted many key proposals on topics such as values, variables, pointers, references, and inheritance. Here is some sample code illustrating some of these features[[1]](https://github.com/carbon-language/carbon-lang/tree/trunk/toolchain/check/testdata/class/base.carbon):

```
base class Base {
  var b: i32;
}

class Derived {
  extend base: Base;

  var d: i32;
}

fn Make() -> Derived {
  return {.base = {.b = 4}, .d = 7};
}
```

We also engaged more deeply with the C++ community, but it's clear we need a working compiler to further the conversation. This need for tooling is driving our 2024 priorities, with a potential v0.1 release in 2025. Get more details on our deliverables in the announcements below.

To see what v0.1 might include, check out "[Defining the 0.1 language](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2759.md)" and expected [milestones](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/milestones.md). 

## **Top announcements**

Announcement: [2024 roadmap](https://github.com/carbon-language/carbon-lang/blob/e5d5acc035c6e077e4bf63c557ea6626890372f8/docs/project/roadmap.md) from [#3564](https://github.com/carbon-language/carbon-lang/pull/3564)

*   Objective for 2024: a working toolchain that supports C++ interop
    *   Building realistic Carbon code for interesting interop with C++
    *   Building realistic C++ code for interesting interop with Carbon
    *   The interop itself allows a single program mixing the two languages
*   Key results for 2024
    *   Carbon's toolchain implements enough of the language to build realistic code
    *   Carbon's toolchain can build C++ code
    *   Carbon's toolchain works with existing, simple C++ build systems
    *   Carbon has a design and toolchain implementation of basic C++ interop
    *   Give talks at 2-3 conferences covering 3-4 different Carbon topics

Announcement: [Focus implementation effort on the toolchain](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p3532.md) [#3532](https://github.com/carbon-language/carbon-lang/pull/3532)

## **Talks and blogs from the team in 2023**

Talks:

*   [Carbon’s Successor Strategy: From C++ interop to memory safety](https://www.youtube.com/watch?v=1ZTJ9omXOQ0), C++Now
*   Definition-Checked Generics ([Part 1](https://www.youtube.com/watch?v=FKC8WACSMP0), [Part 2](https://www.youtube.com/watch?v=VxQ3PwxiSzk)), C++Now
*   [Modernizing Compiler Design for Carbon’s Toolchain](https://www.youtube.com/watch?v=ZI198eFghJk), C++Now

Recent Carbon blog posts

*   [The little things #1: Access control - Coding in Old Entish](https://chandlerc.blog/posts/little_things_1_access/)
*   [Improving the user-facing Carbon Explorer output](https://chandlerc.blog/posts/2023/11/improving_carbon_explorer_output/)
*   [Summer of Code: Editor Integration](https://chandlerc.blog/posts/2023/11/editor_integration_gsoc/)

## **<span style="text-decoration:underline;">Top proposals and issues from 2023</span>**

*   <In progress> Variadics [#2240](https://github.com/carbon-language/carbon-lang/pull/2240)
*   Ranged-based for for user-defined types [#1885](https://github.com/carbon-language/carbon-lang/pull/1885)
*   Character literals [#1964](https://github.com/carbon-language/carbon-lang/pull/1964)
*   Values, variables, pointers, and references [#2006](https://github.com/carbon-language/carbon-lang/pull/2006)
*   [Associated constant assignment versus equality](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2173.md) [#2173](https://github.com/carbon-language/carbon-lang/pull/2173)
*   [Pattern matching syntax and semantics](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2188.md) [#2188](https://github.com/carbon-language/carbon-lang/pull/2188)
*   [Assignment statements](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2511.md) [#2511](https://github.com/carbon-language/carbon-lang/pull/2511)
*   Simplified package declaration for the `Main` package [#2550](https://github.com/carbon-language/carbon-lang/pull/2550)
*   Termination algorithm for impl selection [#2687](https://github.com/carbon-language/carbon-lang/pull/2687)
*   Consistent `class` and `interface` syntax [#2760](https://github.com/carbon-language/carbon-lang/pull/2760)
*   Functions, function types, and function calls [#2875](https://github.com/carbon-language/carbon-lang/pull/2875)
*   Checked generics calling templates [#2153](https://github.com/carbon-language/carbon-lang/issues/2153)
*   Name conflicts between base and derived classes [#2355](https://github.com/carbon-language/carbon-lang/issues/2355)

## **Other notes**

If you like this set of links, you'll love the [weekly meeting notes](https://docs.google.com/document/d/1xEwSnJoYQlDFUL9JH6ALDN5ihonGd_IxfIIyyqzz3wg/edit), from the Carbon Weekly Sync.

## **Wrap-up**

Don't forget to subscribe; you can join [announce@carbon-lang.dev](https://groups.google.com/a/carbon-lang.dev/g/announce). If you have comments or would like to contribute to future editions of **Carbon Copy**, please reach out.  And, always, [join us](https://github.com/carbon-language/carbon-lang/blob/trunk/README.md#join-us) any way you can!

**Comment by Nv7-GitHub** [comment]
The `Derived` reminds me a lot of Go

**Comment by garrdbyrd** [comment]
Can a class be declared without the `base` keyword?

**Reply by chandlerc** [reply]
Yep. It's a final class in that case. All the details here: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#inheritance

**Comment by tawaliou** [comment]
Hi. Regarding the tooling, I'm curious about Carbon's package management capabilities. Specifically, will Carbon have a package manager and a packages repository similar to npm for JavaScript? Given its compatibility with C++, could this feature be complex to implement ?

**Reply by chandlerc** [reply]
Yes, we'd like to have support for package management: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#language-tools-and-ecosystem:~:text=Infrastructure%20to%20enable%20package%20management

Supporting C++ there will definitely be a challenge, but I'm pretty excited about it. That said, we're still at a very early stage, so it'll be a while before we get into this part of the project.
